emacs/.emacs.d/config.org
#+TITLE: Ryan's Emacs config
#+AUTHOR: Ryan Kes
#+DESCRIPTION: There are many Emacs configs, this one is mine

* About
  My personal Emacs config.
* init.el
   To load this config add the following to ~init.el~

#+begin_example
(org-babel-load-file
 (expand-file-name
  "config.org"
  user-emacs-directory))
#+end_example

* Package Management
** Archives
#+begin_src emacs-lisp
  (setq package-archives '(("melpa" . "https://melpa.org/packages/")
			   ("org" . "https://orgmode.org/elpa/")
			   ("elpa" . "https://elpa.gnu.org/packages/")))

  (package-initialize)
  (unless package-archive-contents
  (package-refresh-contents))
#+end_src
** Use-Package
   The ~use-package~ macro allows you to isolate package configuration in your ~.emacs~ file in a way that is both performance-oriented and, well, tidy.

#+begin_src emacs-lisp
(unless (package-installed-p 'use-package)
  (package-install 'use-package))

(require 'use-package)
  ;; "ensure" packages by default
  (setq use-package-always-ensure t)
#+end_src
** Ensure
   The ~:ensure~ keyword causes the package(s) to be installed automatically if not already present on your system

#+begin_src emacs-lisp
(require 'use-package-ensure)
(setq use-package-always-ensure t)
#+end_src

* General Configuration
** User Interface
*** Start with minimal interface
#+begin_src emacs-lisp
(setq inhibit-startup-message t) ;; Disable startup message

(scroll-bar-mode -1)        ; Disable visible scrollbar
(tool-bar-mode -1)          ; Disable the toolbar
(tooltip-mode -1)           ; Disable tooltips
(menu-bar-mode -1)          ; Disable the menu bar

;; Set up the visible bell
(setq visible-bell t)
#+end_src
*** Enable line numbers and customize their format.
#+begin_src emacs-lisp
  (column-number-mode)
  (global-display-line-numbers-mode t)
  (setq display-line-numbers-type `relative)
#+end_src

** Theme
#+begin_src emacs-lisp
(use-package doom-themes)
(load-theme 'doom-molokai t)
#+end_src

** Font
#+begin_src emacs-lisp
  ;; Default pitch face
  (set-face-attribute 'default nil :font "JetBrainsMono Nerd Font")

  ;; Fixed pitch face
  (set-face-attribute 'fixed-pitch  nil :font "JetBrainsMono Nerd Font")

  ;; Variable pitch face
  (set-face-attribute 'variable-pitch  nil :font "Cantarell")
#+end_src

** Icons
#+begin_src emacs-lisp
(use-package all-the-icons)
#+end_src
  
** Mode Line
*** Doom Modeline
#+begin_src emacs-lisp
    (use-package doom-modeline
  :init (doom-modeline-mode 1)
  :custom (doom-modeline-height 15))
#+end_src


** Escape Cancels All
#+begin_src emacs-lisp
(global-set-key (kbd "<escape>") 'keyboard-escape-quit)
#+end_src 

** Evil Mode
*** General
#+begin_src emacs-lisp
  (use-package evil
    :init      ;; tweak evil's configuration before loading it
    (setq evil-want-keybinding nil)
(setq evil-want-C-u-scroll t)
    (evil-mode))
#+end_src

*** Evil Collection
    Evil Collection is also installed since it adds 'evil' bindings to parts of Emacs that the standard Evil package does not cover, such as: calendar, help-mode and ibuffer.

#+begin_src emacs-lisp
  (use-package evil-collection
    :after evil
    :config
    (evil-collection-init))
#+end_src

** General Keybindings
   [[https://github.com/noctuid/general.el][general.el]] provides a more convenient way to bind keys in emacs for both evil and non-evil users. ~general-define-key~ allows defining multiple keys at once, implicitly wrapping key strings with (kbd ...), having named prefix key sequences (like the leader key in vim), and more.

#+begin_src emacs-lisp
   (use-package general
     :config
     (general-evil-setup t)

  (general-create-definer dw/leader-key-def
     :keymaps '(normal insert visual emacs)
     :prefix "SPC"
     :global-prefix "C-SPC"))
#+end_src

** Keybinding Panel (which-key)
   [[https://github.com/justbur/emacs-which-key][which-key]] shows an overview of what keybindings are available based on the prefix keys you entered.

#+begin_src emacs-lisp
  (use-package which-key
    :init (which-key-mode)
  :config(setq which-key-idle-delay 0.3))
#+end_src

* Ivy
** General
#+begin_src emacs-lisp
    (use-package ivy
      :diminish
      :bind (("C-s" . swiper)
	     :map ivy-minibuffer-map
	     ("TAB" . ivy-alt-done)
	     ("C-l" . ivy-alt-done)
	     ("C-j" . ivy-next-line)
	     ("C-k" . ivy-previous-line)
	     :map ivy-switch-buffer-map
	     ("C-k" . ivy-previous-line)
	     ("C-l" . ivy-done)
	     ("C-d" . ivy-switch-buffer-kill)
	     :map ivy-reverse-i-search-map
	     ("C-k" . ivy-previous-line)
	     ("C-d" . ivy-reverse-i-search-kill))
      :init
      (ivy-mode 1)) ;; Load keybindings
#+end_src

** Ivy-rich
   [[https://github.com/Yevgnen/ivy-rich][ivy-rich]] comes with rich transformers for commands from ~ivy~ and ~counsel~.

#+begin_src emacs-lisp
      (use-package ivy-rich
:after counsel
    :init
  (ivy-rich-mode 1))
#+end_src

** Counsel
   ~ivy-mode~ ensures that any Emacs command using ~completing-read-function~ uses ivy for completion. Counsel takes this further, providing versions of common Emacs commands that are customised to make the best use of Ivy. For example, ~counsel-find-file~ has some additional keybindings. Pressing ~DEL~ will move you to the parent directory.

#+begin_src emacs-lisp
(use-package counsel
  :demand t
  :bind (("M-x" . counsel-M-x)
         ("C-x b" . counsel-ibuffer)
         ("C-x C-f" . counsel-find-file)
         ;; ("C-M-j" . counsel-switch-buffer)
         ("C-M-l" . counsel-imenu)
         :map minibuffer-local-map
         ("C-r" . 'counsel-minibuffer-history))
  :custom
  (counsel-linux-app-format-function #'counsel-linux-app-format-function-name-only)
  :config
  (setq ivy-initial-inputs-alist nil)) ;; Don't start searches with ^
#+end_src

* Development
** Languages
*** Emacs Lisp
**** Helpful
     [[https://github.com/Wilfred/helpful][Helpful]] is an alternative to the built-in Emacs help that provides much more contextual information.

#+begin_src emacs-lisp
  (use-package helpful
    :custom
    (counsel-describe-function-function #'helpful-callable)
    (counsel-describe-variable-function #'helpful-variable)
    :bind
    ([remap describe-function] . helpful-function)
    ([remap describe-symbol] . helpful-symbol)
    ([remap describe-variable] . helpful-variable)
    ([remap describe-command] . helpful-command)
    ([remap describe-key] . helpful-key))
#+end_src

** Productivity
*** Rainbow Delimiters
#+begin_src emacs-lisp
  (use-package rainbow-delimiters
    :hook (prog-mode . rainbow-delimiters-mode))
#+end_src

** Projectile
   [[https://github.com/bbatsov/projectile][Projectile]] is a project interaction library for Emacs. Its goal is to provide a nice set of features operating on a project level without introducing external dependencies (when feasible). For instance - finding project files has a portable implementation written in pure Emacs Lisp without the use of GNU ~find~ (but for performance sake an indexing mechanism backed by external commands exists as well).

#+begin_src emacs-lisp
(use-package projectile
  :ensure t
  :init
  (projectile-mode +1)
  (setq projectile-project-search-path '("~/devel/" "~/dotfiles/" "~/Documents/"))
  :bind (:map projectile-mode-map
              ("s-p" . projectile-command-map)
              ("C-c p" . projectile-command-map)))
#+end_src
*** Counsel-projectile
    [[https://github.com/bbatsov/projectile][Projectile]] has native support for using [[https://github.com/abo-abo/swiper][ivy]] as its completion system. [[https://github.com/ericdanan/counsel-projectile][Counsel-projectile]] provides further ivy integration into projectile by taking advantage of ivy's support for selecting from a list of actions and applying an action without leaving the completion session. Concretely, counsel-projectile defines replacements for existing projectile commands as well as new commands that have no projectile counterparts. A minor mode is also provided that adds key bindings for all these commands on top of the projectile key bindings.

#+begin_src emacs-lisp
  (use-package counsel-projectile
  :after projectile
  :config
  (counsel-projectile-mode))
#+end_src

** Git
*** Magit
    [[https://magit.vc/][Magit]] is a complete text-based user interface to Git. It fills the glaring gap between the Git command-line interface and various GUIs, letting you perform trivial as well as elaborate version control tasks with just a couple of mnemonic key presses. 

#+begin_src emacs-lisp
(use-package magit)
#+end_src

* Org Mode
  [[https://orgmode.org/][Org]] is a highly flexible structured plain text file format, composed of a few simple, yet versatile, structures — constructed to be both simple enough for the novice and powerful enough for the expert. 

#+begin_src emacs-lisp
    (defun efs/org-mode-setup ()
      (org-indent-mode)
      (variable-pitch-mode 1)
      (visual-line-mode 1))

      (use-package org
    :hook
  (org-mode . efs/org-mode-setup)
  :config
    (setq org-ellipsis " ▾"))
#+end_src

** Heading Bullets
[[https://github.com/sabof/org-bullets][org-bullets]] shows org-mode bullets as UTF-8 characters.

#+begin_src emacs-lisp
      (use-package org-bullets
  :hook
(org-bullets-mode t))
#+end_src

** Better Font Faces

#+begin_src emacs-lisp
  (dolist (face '((org-level-1 . 1.2)
                  (org-level-2 . 1.1)
                  (org-level-3 . 1.05)
                  (org-level-4 . 1.0)
                  (org-level-5 . 1.1)
                  (org-level-6 . 1.1)
                  (org-level-7 . 1.1)
                  (org-level-8 . 1.1)))
    (set-face-attribute (car face) nil :font "Cantarell" :weight 'regular :height (cdr face)))

  ;; Ensure that anything that should be fixed-pitch in Org files appears that way
  (set-face-attribute 'org-block nil    :foreground nil :inherit 'fixed-pitch)
  (set-face-attribute 'org-table nil    :inherit 'fixed-pitch)
  (set-face-attribute 'org-formula nil  :inherit 'fixed-pitch)
  (set-face-attribute 'org-code nil     :inherit '(shadow fixed-pitch))
  (set-face-attribute 'org-table nil    :inherit '(shadow fixed-pitch))
  (set-face-attribute 'org-verbatim nil :inherit '(shadow fixed-pitch))
  (set-face-attribute 'org-special-keyword nil :inherit '(font-lock-comment-face fixed-pitch))
  (set-face-attribute 'org-meta-line nil :inherit '(font-lock-comment-face fixed-pitch))
  (set-face-attribute 'org-checkbox nil  :inherit 'fixed-pitch)
  (set-face-attribute 'line-number nil :inherit 'fixed-pitch)
  (set-face-attribute 'line-number-current-line nil :inherit 'fixed-pitch)
#+end_src

** Center Org Buffers
[[https://github.com/joostkremers/visual-fill-column][visual-fill-column-mode]] is a small Emacs minor mode that mimics the effect of ~fill-column~ in ~visual-line-mode~. Instead of wrapping lines at the window edge, which is the standard behaviour of ~visual-line-mode~, it wraps lines at ~fill-column~. If ~fill-column~ is too large for the window, the text is wrapped at the window edge.

#+begin_src emacs-lisp
  (defun efs/org-mode-visual-fill ()
    (setq visual-fill-column-width 120
          visual-fill-column-center-text t)
    (visual-fill-column-mode 1))

  (use-package visual-fill-column
    :hook (org-mode . efs/org-mode-visual-fill))
#+end_src

** Org Agenda
#+begin_src emacs-lisp
    (setq
  org-agenda-files (directory-files-recursively "~/Documents/org/" "\.org$"))
#+end_src


emacs/.config/doom/config.org
* Table of Contents :TOC_3:noexport:
- [[#general-settings][General settings]]
  - [[#variable-font-sizes-depending-on-screen-width-set-font-to-jetbrains-mono][Variable font sizes depending on screen width. Set font to JetBrains Mono.]]
  - [[#theme-is-set-to-doom-molokai-from-doom-themes][Theme is set to doom-molokai from doom-themes]]
  - [[#fetch-password-from-authinfo][Fetch password from authinfo]]
  - [[#set-relative-line-numbers][Set relative line numbers]]
- [[#projectile][Projectile]]
  - [[#set-default-paths-projectile-should-search][Set default paths projectile should search]]
- [[#circe][Circe]]
  - [[#connect-to-znc-passwords-are-gotten-from-password-store][Connect to ZNC. Passwords are gotten from Password Store.]]
  - [[#deft][Deft]]
    - [[#deft-should-use-org-roam-directory][Deft should use org-roam directory]]
- [[#org-mode][Org mode]]
  - [[#general-settings-1][General settings]]
    - [[#when-moving-entries-up-and-down-in-the-buffer-use-j-and-k-instead-of--and-][When moving entries up and down in the buffer use `j` and `k` instead of `↑` and `↓`]]
    - [[#auto-save-org-mode-files][Auto save org-mode files]]
  - [[#org-protocol-capture-html][org-protocol-capture-html]]
    - [[#require-package][Require package]]
  - [[#org-capture-templates][org-capture-templates]]
    - [[#capture-web-data-for-org-protocol-capture-html][Capture web data for org-protocol-capture-html]]
  - [[#org-roam][org-roam]]
    - [[#set-directory][Set directory]]
    - [[#set-org-roam-viewer][Set org-roam-viewer]]
  - [[#org-journal][org-journal]]
    - [[#get-org-journal-to-work-nicely-with-org-roam][Get org-journal to work nicely with org-roam]]
  - [[#publishing][Publishing]]
- [[#mu4e][mu4e]]
  - [[#setup-mu4e-with-multiple-profiles][Setup mu4e with multiple profiles]]
- [[#elfeed][Elfeed]]
  - [[#elfeed-and-elfeed-protocol-to-read-feeds][elfeed and elfeed-protocol to read feeds.]]
  - [[#elfeed-org][elfeed-org]]
  - [[#elfeed-goodies][elfeed-goodies]]
- [[#spelling][Spelling]]
  -  [[#set-dictionaries-for-languages-i-use][Set dictionaries for languages I use]]
- [[#markdown][Markdown]]
  - [[#open-mdx-files-in-markdown-mode][Open mdx files in markdown-mode]]

* General settings
** Variable font sizes depending on screen width. Set font to JetBrains Mono.
#+BEGIN_SRC emacs-lisp
(defun set-font (font-size bigfont-size)
  (setq doom-font (font-spec :family "JetBrainsMono Nerd Font" :size font-size))
  (setq doom-big-font (font-spec :family "JetBrainsMono Nerd Font" :size bigfont-size))
  (setq doom-variable-pitch-font (font-spec :family "Noto Sans" :size font-size))
  (setq doom-serif-font (font-spec :family "Noto Serif" :size font-size))
  (setq doom-unicode-font (font-spec :family "JoyPixels"))
  )

(when window-system
  (if (> (x-display-pixel-width) 2560)
      (set-font 12 24)
    (set-font 16 28)))
#+END_SRC
** Theme is set to doom-molokai from [[https://github.com/hlissner/emacs-doom-themes][doom-themes]]
#+BEGIN_SRC emacs-lisp
(setq doom-theme 'doom-molokai)
#+END_SRC
** Fetch password from authinfo
#+BEGIN_SRC emacs-lisp
(defun my-fetch-password (&rest params)
  (require 'auth-source)
  (let ((match (car (apply #'auth-source-search params))))
    (if match
        (let ((secret (plist-get match :secret)))
          (if (functionp secret)
              (funcall secret)
            secret))
      (error "Password not found for %S" params))))
#+END_SRC
** Set relative line numbers
#+BEGIN_SRC emacs-lisp
(setq display-line-numbers-type `relative)
#+END_SRC
* [[https://www.projectile.mx/en/latest/][Projectile]]
** Set default paths [[https://www.projectile.mx/en/latest/][projectile]] should search
#+BEGIN_SRC emacs-lisp
(use-package! projectile
  :defer t
  :config
  (setq projectile-project-search-path '("~/devel/personal/" "~/devel/andthensome/" "~/Documents/" "~/dotfiles" "~/private-dotfiles", "~/devel/slip-box/", "~/devel/andthensome/clients/"))) ; Default paths
#+END_SRC
* [[https://github.com/jorgenschaefer/circe][Circe]]
** Connect to [[https://wiki.znc.in/ZNC][ZNC]]. Passwords are gotten from [[https://www.passwordstore.org/][Password Store]].
#+BEGIN_SRC emacs-lisp
(use-package! circe
  :defer t
  :config
  (setq circe-network-options
        `(
          ("znc-freenode"
           :host "irc.higherlearning.eu"
           :port "5000"
           :tls t ;; Enable tls
           :user "l0rd/freenode"
           :server-buffer-name "⇄ Freenode (ZNC)"
           :pass (lambda (&rest _) (+pass-get-secret "controlpanel/irc.higherlearning.eu")) ;; Get password from pass
           :channels ("#emacs"))
          ("znc-snoonet"
           :host "irc.higherlearning.eu"
           :port "5000"
           :tls t
           :user "l0rd/snoonet"
           :server-buffer-name "⇄ Snoonet (ZNC)"
           :pass (lambda (&rest _) (+pass-get-secret "controlpanel/irc.higherlearning.eu"))
           :channels ("#islam"))
          ))
  )
(add-hook 'circe-channel-mode-hook 'enable-lui-autopaste)
#+END_SRC

** [[https://jblevins.org/projects/deft/][Deft]]
*** Deft should use org-roam directory
#+BEGIN_SRC emacs-lisp
(setq deft-directory "~/Documents/slip-box")
#+END_SRC
* [[https://orgmode.org/][Org mode]]
** General settings
*** When moving entries up and down in the buffer use `j` and `k` instead of `↑` and `↓`
#+BEGIN_SRC emacs-lisp
(after! org
  (map! :map org-mode-map
        :n "M-j" #'org-metadown
        :n "M-k" #'org-metaup)
  )
#+END_SRC
*** Auto save org-mode files
#+BEGIN_SRC emacs-lisp
(add-hook 'auto-save-hook 'org-save-all-org-buffers)
#+END_SRC

- Set default org directories
- When task is set to `DONE` [[https://orgmode.org/manual/Closing-items.html][add timestamp and give the opportunity to type in a note]]
- Show images in buffers as default
#+BEGIN_SRC emacs-lisp
(setq org-directory "~/Documents/org"
      org-agenda-files (directory-files-recursively "~/Documents/org/" "\.org$")
      org-log-done 'note
      org-default-notes-file (expand-file-name "inbox.org" org-directory)
      org-startup-with-inline-images t)
#+END_SRC
** [[https://github.com/alphapapa/org-protocol-capture-html][org-protocol-capture-html]]
*** Require package
#+BEGIN_SRC emacs-lisp
(use-package! org-protocol-capture-html :defer t)
#+END_SRC
** org-capture-templates
*** Capture web data for [[https://github.com/alphapapa/org-protocol-capture-html][org-protocol-capture-html]]
#+BEGIN_SRC emacs-lisp
(after! org
  (add-to-list 'org-capture-templates '("w" "Web site" entry
                                        (file+olp "~/Documents/org/web.org" "Web")
                                        "* %c\n%U %?%:initial"))
  )
#+END_SRC
** [[https://github.com/org-roam/org-roam][org-roam]]
*** Set directory
#+BEGIN_SRC emacs-lisp
(setq org-roam-directory "~/Documents/slip-box")
#+END_SRC
*** Set org-roam-viewer
#+BEGIN_SRC emacs-lisp
(setq org-roam-graph-viewer "~/.local/bin/viewDotfile")
#+END_SRC
** [[https://github.com/bastibe/org-journal][org-journal]]
*** Get org-journal to work nicely with org-roam
#+BEGIN_SRC emacs-lisp
(setq org-journal-dir "~/Documents/slip-box")
(setq org-journal-date-prefix "#+title: ")
(setq org-journal-file-format "%Y-%m-%d.org")
(setq org-journal-date-format "%A, %d %B %Y")
#+END_SRC
** Publishing
#+BEGIN_SRC emacs-lisp
(setq org-publish-project-alist
      '(("wiki"
         :base-directory "~/Documents/slip-box/"
         :publishing-directory "~/devel/personal/Ryankes.eu/wiki/content/"
         :publishing-function org-pandoc-export-to-markdown_mmd
         :select-tags ("export-blog")
         )))
#+END_SRC
* [[https://www.djcbsoftware.nl/code/mu/mu4e.html][mu4e]]
** Setup mu4e with multiple profiles
#+BEGIN_SRC emacs-lisp
(use-package! mu4e
  :defer t
  :config
  (setq sendmail-program "/usr/bin/msmtp"
        send-mail-function
        'smtpmail-send-it
        message-sendmail-f-is-evil
        t
        message-sendmail-extra-arguments
        '("--read-envelope-from")
        message-send-mail-function
        'message-send-mail-with-sendmail)
  (setq mu4e-maildirs-extension-custom-list '("/ryankes/INBOX" "/ryankes/Sent" "/ryankes/Drafts"
                                              "/ryankes/Archive" "/ryankes/Archive.quicknet"
                                              "/gmail/INBOX" "/gmail/Sent" "/gmail/Drafts"
                                              "/gmail/Archive" "/andthensome/INBOX" "/andthensome/Sent"
                                              "/andthensome/Drafts" "/andthensome/Archive")))
(setq mu4e-maildir "~/.local/share/mail"
      mu4e-maildir-shortcuts
      '(("/ryankes/INBOX" . ?r)
        ("/gmail/INBOX" . ?g)
        ("/andthensome/INBOX" . ?a))
      mu4e-mu-home
      "~/.cache/mu"
      mu4e-update-interval
      300
      mu4e-get-mail-command
      "mailsync"
      mu4e-enable-mode-line
      t
      mu4e-use-fancy-chars
      t
      mu4e-context-policy
      'pick-first
      mu4e-attachment-dir
      "~/Downloads/")
(after! mu4e
  ;; load package to be able to capture emails for GTD
  (require 'org-mu4e)

  (setq mu4e-contexts
        `(,(make-mu4e-context :name "Ryankes"
                              :match-func (lambda (msg)
                                            (when msg
                                              (string-match-p "^/ryankes"
                                                              (mu4e-message-field msg :maildir)))):vars'((user-mail-address . "ryan@ryankes.eu")
                                            (user-full-name . "Ryan")
                                            (mu4e-sent-folder . "/ryankes/Sent")
                                            (mu4e-drafts-folder . "/ryankes/Drafts")
                                            (mu4e-trash-folder . "/ryankes/Trash")
                                            (mu4e-refile-folder . "/ryankes/Archive")))
          ,(make-mu4e-context :name "Gmail"
                              :match-func (lambda (msg)
                                            (when msg
                                              (string-match-p "^/gmail"
                                                              (mu4e-message-field msg :maildir)))):vars'((user-mail-address . "alrayyes@gmail.com")
                                            (user-full-name . "Ryan")
                                            (mu4e-sent-folder . "/gmail/Sent")
                                            (mu4e-drafts-folder . "/gmail/Drafts")
                                            (mu4e-trash-folder . "/gmail/Trash")
                                            (mu4e-refile-folder . "/gmail/Archive")))
          ,(make-mu4e-context :name "Andthensome"
                              :match-func (lambda (msg)
                                            (when msg
                                              (string-match-p "^/andthensome"
                                                              (mu4e-message-field msg :maildir)))):vars'((user-mail-address . "ryan@andthensome.nl")
                                            (user-full-name . "Ryan Kes")
                                            (mu4e-sent-folder . "/andthensome/Sent")
                                            (mu4e-drafts-folder . "/andthensome/Drafts")
                                            (mu4e-trash-folder . "/andthensome/Trash")
                                            (mu4e-refile-folder . "/andthensome/Archive"))))))
(add-hook 'mu4e-view-mode-hook 'visual-line-mode)
#+END_SRC
* [[https://github.com/skeeto/elfeed][Elfeed]]
** elfeed and [[https://github.com/fasheng/elfeed-protocol][elfeed-protocol]] to read feeds.
#+BEGIN_SRC emacs-lisp
(use-package! elfeed
  :defer t
  :config
  (setq elfeed-use-curl t)
  (elfeed-set-timeout 36000)
  (setq elfeed-protocol-ttrss-maxsize 200) ; bigger than 200 is invalid
  (elfeed-protocol-enable)
  (defadvice elfeed (after configure-elfeed-feeds activate)
    "Make elfeed-org autotags rules works with elfeed-protocol."
    (defvar elfeed-protocol-tags)
    (setq elfeed-protocol-tags elfeed-feeds)
    (setq elfeed-feeds (list
                        (list "ttrss+https://alrayyes@rss.higherlearning.eu"
                              :password (password-store-get "emacs/rss")
                              :autotags elfeed-protocol-tags)))))
#+END_SRC
** [[https://github.com/remyhonig/elfeed-org][elfeed-org]]
#+BEGIN_SRC emacs-lisp
(use-package! elfeed-org
  :defer t
  :config
  (setq rmh-elfeed-org-files '("~/Documents/org/elfeed.org"))
  )
#+END_SRC
** [[https://github.com/algernon/elfeed-goodies][elfeed-goodies]]
#+BEGIN_SRC emacs-lisp
(use-package! elfeed-goodies
  :defer t
  :config
  (elfeed-goodies/setup))
#+END_SRC
* Spelling
**  Set dictionaries for languages I use
#+BEGIN_SRC emacs-lisp
(with-eval-after-load "ispell"
    (setq ispell-program-name "hunspell")
    ;; ispell-set-spellchecker-params has to be called
    ;; before ispell-hunspell-add-multi-dic will work
    (ispell-set-spellchecker-params)
    (ispell-hunspell-add-multi-dic "en_GB,nl_NL,en_US")
    (setq ispell-dictionary "en_GB,nl_NL,en_US"))
#+END_SRC
* Markdown
** Open mdx files in markdown-mode
#+BEGIN_SRC emacs-lisp
   (add-to-list 'auto-mode-alist '("\\.mdx\\'" . markdown-mode))
#+END_SRC


emacs/.config/doom/config.el
(when window-system
  (if (> (x-display-pixel-width) 2560)
      (setq doom-font (font-spec :family "JetBrains Mono" :size 12))
    (setq doom-font (font-spec :family "JetBrains Mono" :size 16))))

(setq doom-theme 'doom-molokai)

(defun my-fetch-password (&rest params)
  (require 'auth-source)
  (let ((match (car (apply #'auth-source-search params))))
    (if match
        (let ((secret (plist-get match :secret)))
          (if (functionp secret)
              (funcall secret)
            secret))
      (error "Password not found for %S" params))))

(setq display-line-numbers-type `relative)

(use-package! projectile
  :defer t
  :config
  (setq projectile-project-search-path '("~/devel/personal/" "~/devel/andthensome/" "~/Documents/" "~/dotfiles" "~/private-dotfiles", "~/devel/slip-box/"))) ; Default paths

(use-package! circe
  :defer t
  :config
  (setq circe-network-options
        `(
          ("znc-freenode"
           :host "irc.higherlearning.eu"
           :port "5000"
           :tls t ;; Enable tls
           :user "l0rd/freenode"
           :server-buffer-name "⇄ Freenode (ZNC)"
           :pass (lambda (&rest _) (+pass-get-secret "controlpanel/irc.higherlearning.eu")) ;; Get password from pass
           :channels ("#emacs"))
          ("znc-snoonet"
           :host "irc.higherlearning.eu"
           :port "5000"
           :tls t
           :user "l0rd/snoonet"
           :server-buffer-name "⇄ Snoonet (ZNC)"
           :pass (lambda (&rest _) (+pass-get-secret "controlpanel/irc.higherlearning.eu"))
           :channels ("#islam"))
          ))
  )
(add-hook 'circe-channel-mode-hook 'enable-lui-autopaste)

(setq deft-directory "~/Documents/slip-box")

(after! org
  (map! :map org-mode-map
        :n "M-j" #'org-metadown
        :n "M-k" #'org-metaup)
  )

(add-hook 'auto-save-hook 'org-save-all-org-buffers)

(setq org-directory "~/Documents/org"
      org-agenda-files (directory-files-recursively "~/Documents/org/" "\.org$")
      org-log-done 'note
      org-startup-with-inline-images t)

(use-package! org-protocol-capture-html :defer t)

(after! org
  (add-to-list 'org-capture-templates '("w" "Web site" entry
                                        (file+olp "~/Documents/org/web.org" "Web")
                                        "* %c\n%U %?%:initial"))
  )

(setq org-roam-directory "~/Documents/slip-box")

(setq org-roam-graph-viewer "~/.local/bin/viewDotfile")

(setq org-journal-dir "~/Documents/slip-box")
(setq org-journal-date-prefix "#+title: ")
(setq org-journal-file-format "%Y-%m-%d.org")
(setq org-journal-date-format "%A, %d %B %Y")

(setq org-publish-project-alist
      '(("wiki"
         :base-directory "~/Documents/slip-box/"
         :publishing-directory "~/devel/personal/Ryankes.eu/wiki/content/"
         :publishing-function org-pandoc-export-to-markdown_mmd
         :select-tags ("export-blog")
         )))

(use-package! mu4e
  :defer t
  :config
  (setq sendmail-program "/usr/bin/msmtp"
        send-mail-function
        'smtpmail-send-it
        message-sendmail-f-is-evil
        t
        message-sendmail-extra-arguments
        '("--read-envelope-from")
        message-send-mail-function
        'message-send-mail-with-sendmail)
  (setq mu4e-maildirs-extension-custom-list '("/ryankes/INBOX" "/ryankes/Sent" "/ryankes/Drafts"
                                              "/ryankes/Archive" "/ryankes/Archive.quicknet"
                                              "/gmail/INBOX" "/gmail/Sent" "/gmail/Drafts"
                                              "/gmail/Archive" "/andthensome/INBOX" "/andthensome/Sent"
                                              "/andthensome/Drafts" "/andthensome/Archive")))
(setq mu4e-maildir "~/.local/share/mail"
      mu4e-maildir-shortcuts
      '(("/ryankes/INBOX" . ?r)
        ("/gmail/INBOX" . ?g)
        ("/andthensome/INBOX" . ?a))
      mu4e-mu-home
      "~/.cache/mu"
      mu4e-update-interval
      300
      mu4e-get-mail-command
      "mailsync"
      mu4e-enable-mode-line
      t
      mu4e-use-fancy-chars
      t
      mu4e-context-policy
      'pick-first
      mu4e-attachment-dir
      "~/Downloads/")
(after! mu4e
  ;; load package to be able to capture emails for GTD
  (require 'org-mu4e)

  (setq mu4e-contexts
        `(,(make-mu4e-context :name "Ryankes"
                              :match-func (lambda (msg)
                                            (when msg
                                              (string-match-p "^/ryankes"
                                                              (mu4e-message-field msg :maildir)))):vars'((user-mail-address . "ryan@ryankes.eu")
                                            (user-full-name . "Ryan")
                                            (mu4e-sent-folder . "/ryankes/Sent")
                                            (mu4e-drafts-folder . "/ryankes/Drafts")
                                            (mu4e-trash-folder . "/ryankes/Trash")
                                            (mu4e-refile-folder . "/ryankes/Archive")))
          ,(make-mu4e-context :name "Gmail"
                              :match-func (lambda (msg)
                                            (when msg
                                              (string-match-p "^/gmail"
                                                              (mu4e-message-field msg :maildir)))):vars'((user-mail-address . "alrayyes@gmail.com")
                                            (user-full-name . "Ryan")
                                            (mu4e-sent-folder . "/gmail/Sent")
                                            (mu4e-drafts-folder . "/gmail/Drafts")
                                            (mu4e-trash-folder . "/gmail/Trash")
                                            (mu4e-refile-folder . "/gmail/Archive")))
          ,(make-mu4e-context :name "Andthensome"
                              :match-func (lambda (msg)
                                            (when msg
                                              (string-match-p "^/andthensome"
                                                              (mu4e-message-field msg :maildir)))):vars'((user-mail-address . "ryan@andthensome.nl")
                                            (user-full-name . "Ryan Kes")
                                            (mu4e-sent-folder . "/andthensome/Sent")
                                            (mu4e-drafts-folder . "/andthensome/Drafts")
                                            (mu4e-trash-folder . "/andthensome/Trash")
                                            (mu4e-refile-folder . "/andthensome/Archive"))))))
(add-hook 'mu4e-view-mode-hook 'visual-line-mode)

(use-package! elfeed
  :defer t
  :config
  (setq elfeed-use-curl t)
  (elfeed-set-timeout 36000)
  (setq elfeed-protocol-ttrss-maxsize 200) ; bigger than 200 is invalid
  (elfeed-protocol-enable)
  (defadvice elfeed (after configure-elfeed-feeds activate)
    "Make elfeed-org autotags rules works with elfeed-protocol."
    (defvar elfeed-protocol-tags)
    (setq elfeed-protocol-tags elfeed-feeds)
    (setq elfeed-feeds (list
                        (list "ttrss+https://alrayyes@rss.higherlearning.eu"
                              :password (password-store-get "emacs/rss")
                              :autotags elfeed-protocol-tags)))))

(use-package! elfeed-org
  :defer t
  :config
  (setq rmh-elfeed-org-files '("~/Documents/org/elfeed.org"))
  )

(use-package! elfeed-goodies
  :defer t
  :config
  (elfeed-goodies/setup))

(with-eval-after-load "ispell"
    (setq ispell-program-name "hunspell")
    ;; ispell-set-spellchecker-params has to be called
    ;; before ispell-hunspell-add-multi-dic will work
    (ispell-set-spellchecker-params)
    (ispell-hunspell-add-multi-dic "en_GB,nl_NL,en_US")
    (setq ispell-dictionary "en_GB,nl_NL,en_US"))

(add-to-list 'auto-mode-alist '("\\.mdx\\'" . markdown-mode))
